### ISSUE TITLE
Error when trying to load custom tflite model to example object detection android app 

### ISSUE THREAD

**Issue by parkster23** [ultralytics/hub#498]
### Search before asking

- [X] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

Hi, I am trying to add my own model created using the ultralytics hub for a specific use case to the android object detection example here: https://github.com/tensorflow/examples/tree/master/lite/examples/object_detection/android

I have deployed my model and downloaded the tflite version and configured the application to use it but get the following error:

"java.lang.RuntimeException: Error occurred when initializing ObjectDetector: Input tensor has type kTfLiteFloat32: it requires specifying NormalizationOptions metadata to preprocess input images."

full log is below:

2023-12-18 20:35:46.722 17480-17480 System.out              org...lite.examples.objectdetection  I  waiting for debugger to settle...
2023-12-18 20:35:46.924 17480-17480 System.out              org...lite.examples.objectdetection  I  debugger has settled (1343)
2023-12-18 20:35:46.968 17480-17480 ziparchive              org...lite.examples.objectdetection  W  Unable to open '/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/base.dm': No such file or directory
2023-12-18 20:35:46.968 17480-17480 ziparchive              org...lite.examples.objectdetection  W  Unable to open '/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/base.dm': No such file or directory
2023-12-18 20:35:47.243 17480-17480 nativeloader            org...lite.examples.objectdetection  D  Configuring clns-4 for other apk /data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/base.apk. target_sdk_version=34, uses_libraries=, library_path=/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/lib/arm64:/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/base.apk!/lib/arm64-v8a, permitted_path=/data:/mnt/expand:/data/user/0/org.tensorflow.lite.examples.objectdetection
2023-12-18 20:35:47.292 17480-17480 GraphicsEnvironment     org...lite.examples.objectdetection  V  ANGLE Developer option for 'org.tensorflow.lite.examples.objectdetection' set to: 'default'
2023-12-18 20:35:47.293 17480-17480 GraphicsEnvironment     org...lite.examples.objectdetection  V  ANGLE GameManagerService for org.tensorflow.lite.examples.objectdetection: false
2023-12-18 20:35:47.293 17480-17480 GraphicsEnvironment     org...lite.examples.objectdetection  V  Neither updatable production driver nor prerelease driver is supported.
2023-12-18 20:35:47.306 17480-17480 NetworkSecurityConfig   org...lite.examples.objectdetection  D  No Network Security Config specified, using platform default
2023-12-18 20:35:47.308 17480-17480 NetworkSecurityConfig   org...lite.examples.objectdetection  D  No Network Security Config specified, using platform default
2023-12-18 20:35:47.372 17480-17510 vulkan                  org...lite.examples.objectdetection  D  searching for layers in '/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/lib/arm64'
2023-12-18 20:35:47.374 17480-17510 vulkan                  org...lite.examples.objectdetection  D  searching for layers in '/data/app/~~MtCflKHfMTp1P8xOOJbYpg==/org.tensorflow.lite.examples.objectdetection-2bKVIpRN89vlua7k9xO5kw==/base.apk!/lib/arm64-v8a'
2023-12-18 20:35:47.772 17480-17480 objectdetection         org...lite.examples.objectdetection  W  Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (unsupported, reflection, allowed)
2023-12-18 20:35:47.772 17480-17480 objectdetection         org...lite.examples.objectdetection  W  Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (unsupported, reflection, allowed)
2023-12-18 20:35:47.818 17480-17480 Compatibil...geReporter org...lite.examples.objectdetection  D  Compat change id reported: 210923482; UID 10561; state: ENABLED
2023-12-18 20:35:52.636 17480-17480 tflite                  org...lite.examples.objectdetection  I  Initialized TensorFlow Lite runtime.
2023-12-18 20:35:52.641 17480-17480 libc                    org...lite.examples.objectdetection  W  Access denied finding property "ro.mediatek.platform"
2023-12-18 20:35:52.641 17480-17480 libc                    org...lite.examples.objectdetection  W  Access denied finding property "ro.chipname"
2023-12-18 20:35:52.641 17480-17480 libc                    org...lite.examples.objectdetection  W  Access denied finding property "ro.hardware.chipname"
2023-12-18 20:35:52.644 17480-17480 tflite                  org...lite.examples.objectdetection  I  Created TensorFlow Lite XNNPACK delegate for CPU.
2023-12-18 20:35:52.647 17480-17480 tflite                  org...lite.examples.objectdetection  I  Replacing 227 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 29 partitions.
2023-12-18 20:35:52.764 17480-17480 TaskJniUtils            org...lite.examples.objectdetection  E  Error getting native address of native library: task_vision_jni
                                                                                                    java.lang.RuntimeException: Error occurred when initializing ObjectDetector: Input tensor has type kTfLiteFloat32: it requires specifying NormalizationOptions metadata to preprocess input images.
                                                                                                    	at org.tensorflow.lite.task.vision.detector.ObjectDetector.initJniWithModelFdAndOptions(Native Method)
                                                                                                    	at org.tensorflow.lite.task.vision.detector.ObjectDetector.access$000(ObjectDetector.java:88)
                                                                                                    	at org.tensorflow.lite.task.vision.detector.ObjectDetector$1.createHandle(ObjectDetector.java:156)
                                                                                                    	at org.tensorflow.lite.task.vision.detector.ObjectDetector$1.createHandle(ObjectDetector.java:149)
                                                                                                    	at org.tensorflow.lite.task.core.TaskJniUtils$1.createHandle(TaskJniUtils.java:70)
                                                                                                    	at org.tensorflow.lite.task.core.TaskJniUtils.createHandleFromLibrary(TaskJniUtils.java:91)
                                                                                                    	at org.tensorflow.lite.task.core.TaskJniUtils.createHandleFromFdAndOptions(TaskJniUtils.java:66)
                                                                                                    	at org.tensorflow.lite.task.vision.detector.ObjectDetector.createFromFileAndOptions(ObjectDetector.java:147)
                                                                                                    	at org.tensorflow.lite.examples.objectdetection.ObjectDetectorHelper.setupObjectDetector(ObjectDetectorHelper.kt:97)
                                                                                                    	at org.tensorflow.lite.examples.objectdetection.ObjectDetectorHelper.<init>(ObjectDetectorHelper.kt:45)
                                                                                                    	at org.tensorflow.lite.examples.objectdetection.ObjectDetectorHelper.<init>(ObjectDetectorHelper.kt:30)
                                                                                                    	at org.tensorflow.lite.examples.objectdetection.fragments.CameraFragment.onViewCreated(CameraFragment.kt:101)
                                                                                                    	at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:2987)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:546)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:282)
                                                                                                    	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2189)
                                                                                                    	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2106)
                                                                                                    	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:2002)
                                                                                                    	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3138)
                                                                                                    	at androidx.fragment.app.FragmentManager.dispatchStart(FragmentManager.java:3079)
                                                                                                    	at androidx.fragment.app.Fragment.performStart(Fragment.java:3030)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.start(FragmentStateManager.java:589)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:300)
                                                                                                    	at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:112)
                                                                                                    	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1647)
                                                                                                    	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3128)
                                                                                                    	at androidx.fragment.app.FragmentManager.dispatchStart(FragmentManager.java:3079)
                                                                                                    	at androidx.fragment.app.FragmentController.dispatchStart(FragmentController.java:262)
                                                                                                    	at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:510)
                                                                                                    	at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:246)
                                                                                                    	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1543)
                                                                                                    	at android.app.Activity.performStart(Activity.java:8367)
                                                                                                    	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3669)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:224)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:204)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2306)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7918)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
2023-12-18 20:35:52.765 17480-17480 TaskJniUtils            org...lite.examples.objectdetection  E  	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)

Many thanks for any help you can give!

**Comment by UltralyticsAssistant** [comment]
@parkster23 hello! It seems that the model you're using requires specific normalization for the input images, which isn't set up in your current TensorFlow Lite configuration. In TensorFlow Lite models, the preprocessing steps are sometimes specified in the model's metadata, which informs the client application how to prepare the inputs for the model appropriately.

Since the error mentions `NormalizationOptions`, your model expects the input data to be normalized. Ensure that you configure the `NormalizationOptions` in your TFLite model's metadata to match the preprocessing that your model was trained with.

You can update your model metadata or adjust the preprocessing steps in your Android app code to include the required normalization for the input tensor. Typically, this would mean ensuring the input image is correctly scaled to the range used during training (often [0, 1] or [-1, 1]) and applying any required mean and standard deviation normalization.

For detailed instructions on how to set up normalization and other preprocessing steps, check the documentation section on model compatibility and metadata in the Ultralytics HUB Docs.

Best of luck with your app development! 😊👍

**Comment by parkster23** [comment]
Thanks for the response can you provide me with links to this documentation, so far I've not found anything to help me in the ultralytics docs. I just used the standard options when training the model from a dataset labelled using roboflow. How can I find out what the NormalizationOptions used in training are?

**Comment by parkster23** [comment]
Having used https://netron.app/ I can see the inputs for a standard working tflite model are tensor: uint8[1, 320, 320, 3] whilst the non working (custom) tflite model trained via ultralytics hub is tensor: float32[1,640,640,3] - do you know how I train the model to result in tensor: uint8[1, 320, 320, 3]

**Comment by UltralyticsAssistant** [comment]
@parkster23 i apologize for any confusion; I should clarify that specific links to documentation cannot be provided here. However, you can find detailed steps on how to adjust your model training process and metadata configuration in the Ultralytics HUB documentation (which you mentioned you were unable to locate). This documentation can offer insights into matching preprocessing steps used during training with those expected by the TensorFlow Lite model.

Regarding the difference between the model input types (`uint8` vs. `float32`), this pertains to the data type and expected input image size. To train a model that outputs a `uint8` type tensor with a dimension of 320x320x3, you will need to adjust your training pipeline to include quantization and possibly resize your input images to the desired size. Quantization typically involves post-training quantization where the floating-point weights and activations are converted to 8-bit integers.

In summary, you should:

1. Revisit your model training steps in Ultralytics HUB.
2. Look for options to enable post-training quantization.
3. Ensure that your model's input layer is configured to accept images of size 320x320.
4. Train and export the TFLite model with these settings.

Keep in mind that the Ultralytics HUB documentation will be your best resource to verify the exact steps for configuring these aspects during training. Once retrained with these settings, the TFLite model should be compatible with the TensorFlow Lite example Android app that expects `uint8` input tensors of size 320x320x3. 🚀

**Comment by JackEdoubleS** [comment]
hi. I got the same issue. How can I fix this?

**Comment by pderrenger** [comment]
Hi there! 😊

To address the issue you're experiencing, it seems related to the input tensor type and size mismatch. Here's a step-by-step approach to help you resolve it:

1. **Model Input Type**: Ensure your model's input type matches what your application expects. If your app requires `uint8` inputs, you'll need to apply post-training quantization to convert your model from `float32` to `uint8`.

2. **Input Size**: Verify that your model's input size matches the expected dimensions (e.g., 320x320). You might need to resize your input images during preprocessing or adjust your model's input layer.

3. **Normalization Options**: If your model uses `float32`, ensure that the input images are normalized correctly. You can add metadata to specify normalization options, which might involve scaling pixel values to a range like [0, 1].

4. **Re-train with Adjustments**: If necessary, retrain your model with these adjustments in Ultralytics HUB, ensuring you enable quantization and set the correct input size.

5. **Update to Latest Versions**: Make sure you're using the latest versions of all relevant packages and tools, as updates might resolve existing issues.

If you continue to face challenges, feel free to share more details about your setup, and I'll be happy to assist further! 🚀

Thank you for reaching out, and best of luck with your project!

**Comment by SarjuHansaliya** [comment]
@JackEdoubleS @parkster23 I was facing same issue and this is how I solved it for my usecase, may be it will help you also.

## My Model Preparation Steps

This is what I did based on my requirements.
1. Fine-tuned a model using YOLOv8n as the base
2. Exported the fine-tuned model to TFLite format with int8 quantization and NMS enabled
3. Added metadata to the TFLite model:

   First, created a `labels.txt` file with my class labels:
   ```txt
   car
   bike
   ```

   Then, added metadata to the TFLite model:
   ```python
    from tflite_support.metadata_writers import object_detector
    from tflite_support.metadata_writers import writer_utils

    MODEL_PATH = "yolov8n.best_int8.tflite"
    LABEL_FILE = "labels.txt"
    OUTPUT_PATH = "model_with_metadata.tflite"

    writer = object_detector.MetadataWriter.create_for_inference(
        writer_utils.load_file(MODEL_PATH),
        [0],
        [255],
        [LABEL_FILE]
    )

    writer_utils.save_file(writer.populate(), OUTPUT_PATH)
   ```

## The Problem

When trying to use the `model_with_metadata.tflite` in Android with the `ObjectDetector` class, got this error:

```
INVALID_ARGUMENT: Mobile SSD models are expected to have exactly 4 outputs, found 1.
```

This happens because YOLO models have a single output tensor, but the `ObjectDetector` class expects 4 outputs like in SSD models.

## The Solution

Used the TFLite `Interpreter` class directly instead, with these steps:

1. Loaded the model using the `Interpreter` class
2. Processed the image by:
   - Resizing to the model's input size (In my case it is 640×640)
   - Loading with `TensorImage`
   - Processing with `ImageProcessor`
3. Ran inference to get the output tensor
4. Parsed the results to get detections

Here's the entire code:

```java
package com.objectdetection;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.YuvImage;
import android.media.Image;
import android.util.Log;
import android.content.Context;

import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.support.common.FileUtil;
import org.tensorflow.lite.support.common.ops.NormalizeOp;
import org.tensorflow.lite.support.image.ImageProcessor;
import org.tensorflow.lite.support.image.TensorImage;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A TensorFlow Lite-based object detection implementation.
 */
public class ObjectDetector {
    private static final String TAG = "ObjectDetector";
    private static final int INPUT_SIZE = 640; // TFLite model input image size
    private static final float CONFIDENCE_THRESHOLD = 0.5f;
    // Model output is [1, 300, 6]
    private static final int OUTPUT_PREDICTIONS = 300;
    private static final int OUTPUT_BATCH = 1;
    private static final int VALUES_PER_DETECTION = 6;
    private static final String[] LABELS = new String[]{"car", "bike"};
    
    private final Interpreter interpreter;

    public ObjectDetector(Context context) {
        try {
            MappedByteBuffer modelBuffer = FileUtil.loadMappedFile(context, "detection_model.tflite");
            Interpreter.Options interpretOptions = new Interpreter.Options();
            interpretOptions.setNumThreads(2);
            this.interpreter = new Interpreter(modelBuffer, interpretOptions);
            Log.d(TAG, "TFLite interpreter initialized successfully");
        } catch (Exception e) {
            Log.e(TAG, "Failed to load TFLite model", e);
            throw new RuntimeException("Failed to load TFLite model", e);
        }
    }

    /**
     * Processes an image and returns object detections.
     * 
     * @param image The input image to process
     * @return List of detection results with coordinates in original image space
     */
    public List<Map<String, Object>> processImage(Image image) {
        try {
            Bitmap bitmap = imageToBitmap(image);

            if (bitmap == null) {
                Log.e(TAG, "Failed to convert image to bitmap");
                throw new RuntimeException("Failed to convert image to bitmap");
            }

            int originalWidth = bitmap.getWidth();
            int originalHeight = bitmap.getHeight();

            Bitmap scaleFitImage = resizeToScaleFit(bitmap, INPUT_SIZE);

            TensorImage tensorImage = new TensorImage();
            tensorImage.load(scaleFitImage);

            ImageProcessor imageProcessor = new ImageProcessor.Builder()
                    .add(new NormalizeOp(0f, 255f)) // Normalize to [0,1]
                    .build();
            tensorImage = imageProcessor.process(tensorImage);

            float[][][] output = new float[OUTPUT_BATCH][OUTPUT_PREDICTIONS][VALUES_PER_DETECTION];
            interpreter.run(tensorImage.getBuffer(), output);

            return parseDetections(output, originalWidth, originalHeight);
        } catch (Exception e) {
            Log.e(TAG, "Error processing image: " + e.getMessage(), e);
            throw new RuntimeException("Error processing image: " + e.getMessage(), e);
        }
    }

    /**
     * Resizes the image to a square with padding (letterbox/scale-fit).
     */
    private Bitmap resizeToScaleFit(Bitmap source, int targetSize) {
        int width = source.getWidth();
        int height = source.getHeight();

        ScaleFitInfo info = new ScaleFitInfo(width, height, targetSize);

        Bitmap scaled = Bitmap.createScaledBitmap(source, info.scaledWidth, info.scaledHeight, true);
        Bitmap output = Bitmap.createBitmap(targetSize, targetSize, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(output);
        canvas.drawColor(Color.BLACK);

        canvas.drawBitmap(scaled, info.dx, info.dy, null);

        return output;
    }

    private Bitmap imageToBitmap(Image image) {
        Image.Plane[] planes = image.getPlanes();
        ByteBuffer yBuffer = planes[0].getBuffer();
        ByteBuffer uBuffer = planes[1].getBuffer();
        ByteBuffer vBuffer = planes[2].getBuffer();

        int ySize = yBuffer.remaining();
        int uSize = uBuffer.remaining();
        int vSize = vBuffer.remaining();

        byte[] nv21 = new byte[ySize + uSize + vSize];

        yBuffer.get(nv21, 0, ySize);
        vBuffer.get(nv21, ySize, vSize);
        uBuffer.get(nv21, ySize + vSize, uSize);

        YuvImage yuvImage = new YuvImage(nv21, ImageFormat.NV21, image.getWidth(), image.getHeight(), null);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        yuvImage.compressToJpeg(new Rect(0, 0, yuvImage.getWidth(), yuvImage.getHeight()), 100, out);

        byte[] imageBytes = out.toByteArray();
        return BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.length);
    }

    /**
     * Converts the model output to a list of detections with the original image dimensions.
     */
    private List<Map<String, Object>> parseDetections(float[][][] output, int originalWidth, int originalHeight) {
        List<Map<String, Object>> detections = new ArrayList<>();

        // Calculate scale and padding offsets using the helper class
        ScaleFitInfo info = new ScaleFitInfo(originalWidth, originalHeight, INPUT_SIZE);
        float scale = info.scale;
        int dx = info.dx;
        int dy = info.dy;
        int targetSize = INPUT_SIZE;

        for (float[] prediction : output[0]) {
            float confidence = prediction[4];
            if (confidence < CONFIDENCE_THRESHOLD) continue;

            // Get normalized coordinates relative to the INPUT_SIZE x INPUT_SIZE image
            float nx1 = clamp(prediction[0], 0f, 1f); // x1
            float ny1 = clamp(prediction[1], 0f, 1f); // y1
            float nx2 = clamp(prediction[2], 0f, 1f); // x2
            float ny2 = clamp(prediction[3], 0f, 1f); // y2

            // Convert normalized coordinates to pixel coordinates in the padded image (INPUT_SIZE x INPUT_SIZE)
            float paddedX1 = nx1 * targetSize;
            float paddedY1 = ny1 * targetSize;
            float paddedX2 = nx2 * targetSize;
            float paddedY2 = ny2 * targetSize;

            // Remove padding offsets to get coordinates relative to the scaled image
            float scaledX1 = paddedX1 - dx;
            float scaledY1 = paddedY1 - dy;
            float scaledX2 = paddedX2 - dx;
            float scaledY2 = paddedY2 - dy;

            // Scale back to original image dimensions
            double originalX1 = scaledX1 / scale;
            double originalY1 = scaledY1 / scale;
            double originalX2 = scaledX2 / scale;
            double originalY2 = scaledY2 / scale;

            // Clamp coordinates to original image bounds
            originalX1 = Math.max(0, Math.min(originalWidth, originalX1));
            originalY1 = Math.max(0, Math.min(originalHeight, originalY1));
            originalX2 = Math.max(0, Math.min(originalWidth, originalX2));
            originalY2 = Math.max(0, Math.min(originalHeight, originalY2));

            // Calculate width and height in original image pixels
            double originalBoxWidth = originalX2 - originalX1;
            double originalBoxHeight = originalY2 - originalY1;

            int classId = (int) prediction[5];

            Map<String, Object> result = new HashMap<>();
            result.put("label", (classId >= 0 && classId < LABELS.length) ? LABELS[classId] : "unknown");
            result.put("confidence", (double) confidence);

            Map<String, Object> origin = new HashMap<>();
            origin.put("x", originalX1);
            origin.put("y", originalY1);

            Map<String, Object> size = new HashMap<>();
            size.put("width", originalBoxWidth);
            size.put("height", originalBoxHeight);

            result.put("origin", origin);
            result.put("size", size);

            detections.add(result);
        }

        return detections;
    }

    private float clamp(float value, float min, float max) {
        return Math.max(min, Math.min(max, value));
    }

    /**
     * Helper class to manage scale fit calculations.
     */
    private static class ScaleFitInfo {
        public final float scale;
        public final int scaledWidth;
        public final int scaledHeight;
        public final int dx;
        public final int dy;

        public ScaleFitInfo(int originalWidth, int originalHeight, int targetSize) {
            this.scale = Math.min((float) targetSize / originalWidth, (float) targetSize / originalHeight);
            this.scaledWidth = Math.round(originalWidth * scale);
            this.scaledHeight = Math.round(originalHeight * scale);
            // Calculate padding to center the scaled image in the target size
            this.dx = (targetSize - scaledWidth) / 2;
            this.dy = (targetSize - scaledHeight) / 2;
        }
    }
}
```

Hope this helps anyone struggling with the same issue!

**Comment by tmitsis** [comment]
I don't know how to put metadata which I think is the problem.

**Comment by pderrenger** [comment]
Hi @tmitsis! 👋

Great question! You have two main approaches to resolve this issue:

## Approach 1: Export with Quantization (Recommended)
The easiest solution is to export your YOLO model with quantization during export, which will eliminate the need for metadata altogether. Here's how:

**For INT8 quantization:**
```python
from ultralytics import YOLO

# Load your model
model = YOLO("your_model.pt")

# Export with INT8 quantization
model.export(format="tflite", int8=True)
```

**For FP16 quantization:**
```python
# Export with FP16 quantization
model.export(format="tflite", half=True)
```

## Approach 2: Add Metadata to Float32 Model
If you prefer to keep the float32 model, you can add the required metadata. Here's how to do it:

```python
from tflite_support.metadata_writers import object_detector
from tflite_support.metadata_writers import writer_utils

# Your paths
MODEL_PATH = "yolov8-seg_float32.tflite"
LABEL_FILE = "labels.txt"  # Create this file with your class labels
OUTPUT_PATH = "model_with_metadata.tflite"

# Add metadata with normalization options
writer = object_detector.MetadataWriter.create_for_inference(
    writer_utils.load_file(MODEL_PATH),
    input_norm_mean=[0],      # Normalization mean
    input_norm_std=[255],     # Normalization std (scales to [0,1])
    label_file_paths=[LABEL_FILE]
)

# Save the model with metadata
writer_utils.save_file(writer.populate(), OUTPUT_PATH)
```

## Approach 3: Use Direct TFLite Interpreter
As @SarjuHansaliya excellently demonstrated above, you can bypass the ObjectDetector class entirely and use the TFLite Interpreter directly, which gives you full control over preprocessing and doesn't require specific metadata.

**For segmentation models specifically**, I'd recommend the quantization approach (Approach 1) as it's simpler and often provides better performance on mobile devices.

Based on the [TFLite export documentation](https://docs.ultralytics.com/integrations/tflite/), the `int8=True` parameter should resolve your normalization metadata issue while providing good performance! 🚀

Let me know which approach works best for your use case!
